posts/nyc-subway-data-with-kafka-streams-and-jhipster-part-1.md

```diff
@@ -0,0 +1,329 @@
+
+The drawback of Protocol buffers is that you can't use a simple cURL command to visualize the data. You need to generate data access classes using the structure definition and then parse the API's response using a popular language (C++, C#, Go, Java or Python). The compiler can be downloaded [here](https://developers.google.com/protocol-buffers/docs/downloads), if you're on macOS you can run this [Homebrew](https://brew.sh/) command `brew install protobuf`.
+
+# Microservice generation and MTA feed polling
+
+## Microservice generation using JHipster
+
+I will be using [JHipster](http://www.jhipster.tech/) to generate a standard Spring Boot microservice configured with Kafka. If you never used JHipster, I suggest you to take a look at the [video tutorial](https://www.jhipster.tech/video-tutorial/). The microservice will use Kafka as a message broker and will have no database configured.
+
+Here is the `.yo-rc.json` in case you want to generate you own microservice:
+```json
+{
+  "generator-jhipster": {
+    "promptValues": {
+      "packageName": "io.github.falydoor.mtakafka.producer"
+    },
+    "jhipsterVersion": "5.5.0",
+    "applicationType": "microservice",
+    "baseName": "producer",
+    "packageName": "io.github.falydoor.mtakafka.producer",
+    "packageFolder": "io/github/falydoor/mtakafka/producer",
+    "serverPort": "8081",
+    "authenticationType": "jwt",
+    "cacheProvider": "hazelcast",
+    "enableHibernateCache": false,
+    "websocket": false,
+    "databaseType": "no",
+    "devDatabaseType": "no",
+    "prodDatabaseType": "no",
+    "searchEngine": false,
+    "messageBroker": "kafka",
+    "serviceDiscoveryType": "eureka",
+    "buildTool": "maven",
+    "enableSwaggerCodegen": false,
+    "jwtSecretKey": "",
+    "enableTranslation": false,
+    "testFrameworks": [],
+    "jhiPrefix": "jhi",
+    "clientPackageManager": "npm",
+    "skipClient": true,
+    "skipUserManagement": true
+  }
+}
+```
+
+The next step is to generate the Protobuf's classes using the command `protoc` and the [GTFS definition](https://developers.google.com/transit/gtfs-realtime/gtfs-realtime.proto). Save the definition in the resources folder and then run the below command:
+
+```protoc --java_out=src/main/java/ src/main/resources/gtfs-realtime.proto```
+
+You should now have a java class named `GtfsRealtime` that will be used to parse the API's response.
+A Maven dependency is required as well, here is the xml:
+```xml
+<dependency>
+    <groupId>com.google.protobuf</groupId>
+    <artifactId>protobuf-java</artifactId>
+    <version>3.6.1</version>
+</dependency>
+```
+
+## MTA feed polling
+
+The goal is to retrieve all running subways on each lines using the MTA feed API. To achieve that, each feed id will be polled and a custom filter will be applied to only keep running subways. This can be easily done using the `@Scheduled` annotation provided by Spring, see below for the full code:
```

Comment:
```suggestion
The goal is to retrieve all running subways on each line using the MTA feed API. To achieve that, each feed ID will be polled and a custom filter will be applied to only keep running subways. This can be easily done using the `@Scheduled` annotation provided by Spring, see below for the full code:
```